account_invoice_change_currency/models/account_change_currency.py

```diff
@@ -1,32 +1,78 @@
 # Copyright 2017-2018 Vauxoo
 # License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
 from odoo import fields, models, api
+from odoo.tools import float_compare
 
 
 class AccountInvoice(models.Model):
     _inherit = 'account.invoice'
 
+    custom_rate = fields.Float(
+        digits=(12, 6),
+        default=1,
+        help="Set new currency rate to apply on the invoice\n."
+        "This rate will be taken in order to convert amounts between the "
+        "currency on the invoice and last currency")
+
+    @api.model
+    def create(self, values):
+        invoice = super(AccountInvoice, self).create(values)
+        if self.env.context.get('force_rate'):
+            invoice._toggle_forced_rate()
+        invoice._onchange_currency_change_rate()
+        return invoice
+
     @api.multi
     def action_account_change_currency(self):
         track = self.env['mail.tracking.value']
+        precision = self.env['decimal.precision'].precision_get(
+            'Payment Terms')
         today = fields.Date.today()
         for invoice in self.filtered(lambda x: x.state == 'draft'):
             from_currency = invoice.get_last_currency_id()
-            if not from_currency or from_currency == invoice.currency_id:
+            if not from_currency:
                 continue
+            date_invoice = invoice.date_invoice
+            if not date_invoice:
+                date_invoice = today
             ctx = {'company_id': invoice.company_id.id,
-                   'date': invoice.date_invoice or today}
+                   'date': date_invoice}
             currency = invoice.with_context(**ctx).currency_id
-            rate = self.env['res.currency'].with_context(
-                **ctx)._get_conversion_rate(
+            currency_skip = invoice.get_last_currency_id(True)
+            old_rate_currency, old_rate = invoice.get_last_rate()
+            new_rate = invoice.custom_rate
+            if not invoice.custom_rate:
+                new_rate = currency.with_context(**ctx)._get_conversion_rate(
+                    currency_skip, currency, invoice.company_id,
+                    date_invoice) if from_currency else None
+            if (old_rate_currency == currency and new_rate
+                    and old_rate and float_compare(
+                        new_rate, old_rate, precision_digits=precision) == 0):
+                if not invoice.custom_rate or old_rate == new_rate:
+                    continue
+                new_rate = currency.with_context(**ctx)._get_conversion_rate(
```

Comment:
You are computing this same rate from line 45
What about assign to variable and re-use it